File: src/salesnav.rs
use crate::*;

#[derive(Debug, Deserialize)]
pub(crate) struct SalesNavCapture {
    #[serde(rename = "capturedAt")]
    pub(crate) captured_at: Option<String>,
    pub(crate) source: Option<String>,
    #[serde(default)]
    pub(crate) url: Option<String>,
    #[serde(default, rename = "resumeUrl")]
    pub(crate) resume_url: Option<String>,
    #[serde(default)]
    pub(crate) page: Option<SalesNavCapturePage>,
    #[serde(default)]
    pub(crate) pages: Vec<SalesNavCapturePage>,
    #[serde(default, rename = "stateCounts")]
    pub(crate) state_counts: BTreeMap<String, u32>,
    #[serde(default, rename = "rawRowCount")]
    pub(crate) raw_row_count: Option<u32>,
    #[serde(default, rename = "outputRowCount")]
    pub(crate) output_row_count: Option<u32>,
    pub(crate) rows: Vec<SalesNavCaptureRow>,
}

#[derive(Debug, Clone, Deserialize)]
pub(crate) struct SalesNavCapturePage {
    #[serde(default)]
    pub(crate) url: Option<String>,
    #[serde(default, rename = "pageLabel")]
    pub(crate) page_label: Option<String>,
}

#[derive(Debug, Clone, Copy, Default)]
pub(crate) struct ImportCaptureOptions {
    pub(crate) only_connectable: bool,
}

#[derive(Debug, Deserialize)]
pub(crate) struct SalesNavAudit {
    #[serde(alias = "peopleCount")]
    pub(crate) people_count: u32,
    #[serde(default, alias = "recentNames")]
    pub(crate) recent_names: Vec<String>,
}

impl SalesNavAudit {
    pub(crate) fn from_path(path: &PathBuf) -> Result<Self> {
        let raw = fs::read_to_string(path)
            .with_context(|| format!("reading audit {}", path.display()))?;
        serde_json::from_str(&raw).with_context(|| format!("parsing audit {}", path.display()))
    }
}

#[derive(Debug, Deserialize)]
pub(crate) struct SalesNavSendResult {
    pub(crate) candidate: SalesNavSendCandidate,
    pub(crate) status: String,
    #[serde(default)]
    pub(crate) send: Option<serde_json::Value>,
}

impl SalesNavSendResult {
    pub(crate) fn from_path(path: &PathBuf) -> Result<Self> {
        let raw =
            fs::read_to_string(path).with_context(|| format!("reading send {}", path.display()))?;
        serde_json::from_str(&raw).with_context(|| format!("parsing send {}", path.display()))
    }

    pub(crate) fn to_candidate_status(&self) -> (CandidateStatus, String) {
        match self.status.as_str() {
            "pending-verified" => (
                CandidateStatus::Pending,
                "salesnav-send-one verified Connect - Pending".to_string(),
            ),
            "already-pending" => (
                CandidateStatus::AlreadyPending,
                "salesnav-send-one found already pending".to_string(),
            ),
            "email-required" => (
                CandidateStatus::Skipped,
                "salesnav-send-one stopped on email-required invite flow".to_string(),
            ),
            other => {
                let send = self
                    .send
                    .as_ref()
                    .map(|value| value.to_string())
                    .unwrap_or_else(|| "no send detail".to_string());
                (
                    CandidateStatus::Failed,
                    format!("salesnav-send-one status {other}; {send}"),
                )
            }
        }
    }
}

#[derive(Debug, Deserialize)]
pub(crate) struct SalesNavSendCandidate {
    pub(crate) source: String,
    pub(crate) name: String,
    #[serde(alias = "profileUrl")]
    pub(crate) profile_url: Option<String>,
}

impl SalesNavCapture {
    pub(crate) fn from_path(path: &PathBuf) -> Result<Self> {
        let raw = fs::read_to_string(path)
            .with_context(|| format!("reading capture {}", path.display()))?;
        serde_json::from_str(&raw).with_context(|| format!("parsing capture {}", path.display()))
    }
}

#[derive(Debug, Clone, Deserialize)]
pub(crate) struct SalesNavCaptureRow {
    pub(crate) index: u32,
    pub(crate) name: Option<String>,
    #[serde(rename = "profileUrl")]
    pub(crate) profile_url: Option<String>,
    #[serde(rename = "scrollUrn")]
    pub(crate) scroll_urn: Option<String>,
    #[serde(rename = "visibleState")]
    pub(crate) visible_state: Option<serde_json::Value>,
    #[serde(rename = "menuState")]
    pub(crate) menu_state: Option<String>,
    #[serde(rename = "menuLabels")]
    pub(crate) menu_labels: Option<Vec<SalesNavCaptureMenuLabel>>,
    #[serde(rename = "rowHtmlPath")]
    pub(crate) row_html_path: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub(crate) struct SalesNavCaptureMenuLabel {
    pub(crate) text: Option<String>,
    pub(crate) aria: Option<String>,
}

pub(crate) fn capture_state_count(capture: &SalesNavCapture, state: &str) -> u32 {
    capture.state_counts.get(state).copied().unwrap_or_else(|| {
        capture
            .rows
            .iter()
            .filter(|row| row.menu_state.as_deref() == Some(state))
            .count() as u32
    })
}

pub(crate) fn same_observation_identity(
    left: &CandidateObservation,
    right: &CandidateObservation,
) -> bool {
    if let (Some(left_url), Some(right_url)) = (&left.profile_url, &right.profile_url) {
        normalize_linkedin_url(left_url) == normalize_linkedin_url(right_url)
    } else {
        left.source == right.source && left.name == right.name
    }
}

pub(crate) fn candidate_matches_observation(
    candidate: &CandidateEvent,
    observation: &CandidateObservation,
) -> bool {
    if let (Some(candidate_url), Some(observation_url)) =
        (&candidate.profile_url, &observation.profile_url)
    {
        normalize_linkedin_url(candidate_url) == normalize_linkedin_url(observation_url)
    } else {
        candidate.name == observation.name && candidate.source == observation.source
    }
}

pub(crate) fn is_auto_stale_skip(candidate: &CandidateEvent) -> bool {
    candidate.status == CandidateStatus::Skipped
        && candidate
            .note
            .as_deref()
            .is_some_and(|note| note.contains("auto-skipped stale imported candidate"))
}

pub(crate) fn sales_profile_urn_to_lead_url(urn: &str) -> Option<String> {
    let tuple = urn.split_once('(')?.1.strip_suffix(')')?;
    let mut parts = tuple.split(',').map(str::trim);
    let profile_id = parts.next()?;
    let auth_type = parts.next()?;
    let auth_token = parts.next()?;
    if profile_id.is_empty()
        || auth_type.is_empty()
        || auth_token.is_empty()
        || parts.next().is_some()
    {
        return None;
    }
    Some(format!(
        "https://www.linkedin.com/sales/lead/{profile_id},{auth_type},{auth_token}"
    ))
}

pub(crate) fn push_timing(
    run: &mut Run,
    phase: &str,
    source: Option<String>,
    started: Instant,
    detail: Option<String>,
) {
    let duration_ms = started.elapsed().as_millis().min(u128::from(u64::MAX)) as u64;
    run.timings.push(RunTimingEvent {
        at: Local::now(),
        phase: phase.to_string(),
        source,
        duration_ms,
        detail,
    });
    run.mark_updated();
}

pub(crate) fn capture_to_observations(
    source: &str,
    capture: &SalesNavCapture,
    options: ImportCaptureOptions,
) -> Vec<CandidateObservation> {
    capture
        .rows
        .iter()
        .filter_map(|row| {
            let name = row.name.as_ref()?.trim();
            if name.is_empty() {
                return None;
            }
            let menu_state = row.menu_state.as_deref().unwrap_or("unknown").to_string();
            if options.only_connectable && menu_state != "connectable" {
                return None;
            }
            let menu_labels = row
                .menu_labels
                .clone()
                .unwrap_or_default()
                .into_iter()
                .filter_map(|label| label.text.or(label.aria))
                .map(|label| label.trim().to_string())
                .filter(|label| !label.is_empty())
                .collect::<Vec<_>>();
            let profile_url = row.profile_url.clone().or_else(|| {
                row.scroll_urn
                    .as_deref()
                    .and_then(sales_profile_urn_to_lead_url)
            });
            Some(CandidateObservation {
                imported_at: Local::now(),
                captured_at: capture.captured_at.clone(),
                source: source.to_string(),
                index: row.index,
                name: name.to_string(),
                profile_url,
                sales_profile_urn: row.scroll_urn.clone(),
                visible_state: row.visible_state.clone().unwrap_or(serde_json::Value::Null),
                menu_state,
                menu_labels,
                row_html_path: row.row_html_path.clone(),
            })
        })
        .collect()
}

pub(crate) fn update_capture_cursor(run: &mut Run, source: &str, capture: &SalesNavCapture) {
    let last_page = capture.page.as_ref().or_else(|| capture.pages.last());
    let resume_url = capture
        .resume_url
        .clone()
        .or_else(|| capture.url.clone())
        .or_else(|| last_page.and_then(|page| page.url.clone()));
    let captured_pages = if capture.pages.is_empty() {
        u32::from(capture.page.is_some())
    } else {
        capture.pages.len() as u32
    };
    let raw_row_count = capture
        .raw_row_count
        .unwrap_or_else(|| capture.rows.len() as u32);
    let output_row_count = capture
        .output_row_count
        .unwrap_or_else(|| capture.rows.len() as u32);

    run.capture_cursors.insert(
        source.to_string(),
        SourceCaptureCursor {
            source: source.to_string(),
            updated_at: Local::now(),
            captured_at: capture.captured_at.clone(),
            resume_url,
            page_label: last_page.and_then(|page| page.page_label.clone()),
            captured_pages,
            raw_row_count,
            output_row_count,
            connectable_count: capture_state_count(capture, "connectable"),
            already_pending_count: capture_state_count(capture, "already-pending"),
            missing_trigger_count: capture_state_count(capture, "missing-trigger"),
            state_counts: capture.state_counts.clone(),
        },
    );
}

pub(crate) fn import_capture(
    run: &mut Run,
    capture: SalesNavCapture,
    options: ImportCaptureOptions,
) -> Result<usize> {
    let source = capture
        .source
        .clone()
        .or_else(|| run.next_source().map(|next| next.name))
        .ok_or_else(|| anyhow!("capture did not include source and run has no next source"))?;
    ensure_known_source(run, &source)?;
    update_capture_cursor(run, &source, &capture);

    let mut imported = 0;
    for observation in capture_to_observations(&source, &capture, options) {
        let existing_index = run
            .observations
            .iter()
            .position(|existing| same_observation_identity(existing, &observation));
        if let Some(index) = existing_index {
            run.observations[index] = observation;
        } else {
            run.observations.push(observation);
            imported += 1;
        }
    }
    Ok(imported)
}

pub(crate) fn import_capture_into_reservoir(
    reservoir: &mut CandidateReservoir,
    capture: SalesNavCapture,
    options: ImportCaptureOptions,
) -> Result<usize> {
    let source = capture
        .source
        .clone()
        .ok_or_else(|| anyhow!("capture did not include source"))?;
    let mut imported = 0;
    for observation in capture_to_observations(&source, &capture, options) {
        let existing_index = reservoir
            .observations
            .iter()
            .position(|existing| same_observation_identity(existing, &observation));
        if let Some(index) = existing_index {
            reservoir.observations[index] = observation;
        } else {
            reservoir.observations.push(observation);
            imported += 1;
        }
    }
    reservoir.updated_at = Some(Local::now());
    Ok(imported)
}

pub(crate) fn fill_run_from_reservoir(
    run: &mut Run,
    reservoir: &mut CandidateReservoir,
    source: &str,
    limit: usize,
) -> Result<usize> {
    ensure_known_source(run, source)?;
    let mut selected_keys = Vec::new();
    let mut imported = 0;
    for observation in reservoir.available_for_run_source(run, source) {
        if imported >= limit {
            break;
        }
        let mut observation = observation.clone();
        observation.imported_at = Local::now();
        selected_keys.push(observation_key(&observation));
        run.observations.push(observation);
        imported += 1;
    }
    if imported > 0 {
        reservoir
            .observations
            .retain(|observation| !selected_keys.contains(&observation_key(observation)));
        reservoir.updated_at = Some(Local::now());
        run.mark_updated();
    }
    Ok(imported)
}

pub(crate) fn fill_run_from_reservoir_for_top_up(
    run: &mut Run,
    reservoir: &mut CandidateReservoir,
    source: &str,
    limit: usize,
) -> Result<usize> {
    ensure_known_source(run, source)?;
    let mut selected_keys = Vec::new();
    let mut selected = Vec::new();
    for observation in reservoir
        .observations
        .iter()
        .filter(|observation| {
            observation.source == source
                && observation.menu_state == "connectable"
                && !run.has_top_up_blocking_event_for_observation(observation)
                && !run
                    .observations
                    .iter()
                    .any(|existing| same_observation_identity(existing, observation))
        })
        .take(limit)
    {
        let mut observation = observation.clone();
        observation.imported_at = Local::now();
        selected_keys.push(observation_key(&observation));
        selected.push(observation);
    }

    let imported = selected.len();
    if imported > 0 {
        run.observations.extend(selected);
        reservoir
            .observations
            .retain(|observation| !selected_keys.contains(&observation_key(observation)));
        reservoir.updated_at = Some(Local::now());
        run.mark_updated();
    }
    Ok(imported)
}

pub(crate) fn observation_key(
    observation: &CandidateObservation,
) -> (String, String, Option<String>) {
    (
        observation.source.clone(),
        observation.name.clone(),
        observation
            .profile_url
            .as_deref()
            .map(normalize_linkedin_url),
    )
}

pub(crate) fn source_yield_stats(run: &Run, source: &SourcePlan) -> SourceYieldStats {
    let cursor = run.capture_cursors.get(&source.name);
    let raw_row_count = cursor.map(|cursor| cursor.raw_row_count).unwrap_or(0);
    let connectable_count = cursor
        .map(|cursor| cursor.connectable_count)
        .unwrap_or_else(|| {
            run.observations
                .iter()
                .filter(|observation| {
                    observation.source == source.name.as_str()
                        && observation.menu_state == "connectable"
                })
                .count() as u32
        });
    let already_pending_count = cursor
        .map(|cursor| cursor.already_pending_count)
        .unwrap_or_default();
    let email_required_skips = run
        .candidates
        .iter()
        .filter(|candidate| candidate.source == source.name.as_str())
        .filter(|candidate| candidate.status == CandidateStatus::Skipped)
        .filter(|candidate| {
            candidate
                .note
                .as_deref()
                .is_some_and(|note| note.to_ascii_lowercase().contains("email-required"))
        })
        .count() as u32;
    let pending_sends = run.source_verified_count(&source.name);
    let connectable_yield = if raw_row_count > 0 {
        Some(f64::from(connectable_count) / f64::from(raw_row_count))
    } else {
        None
    };
    let recommendation = match connectable_yield {
        Some(yield_rate) if raw_row_count >= 50 && yield_rate <= 0.05 => {
            "low-yield: consider reservoir/fallback before deeper capture".to_string()
        }
        Some(yield_rate) if raw_row_count >= 25 && yield_rate <= 0.10 => {
            "thin: capture with a small buffer and be ready to carry over".to_string()
        }
        Some(_) => "ok".to_string(),
        None => "no capture data".to_string(),
    };
    SourceYieldStats {
        source: source.name.clone(),
        raw_row_count,
        connectable_count,
        already_pending_count,
        email_required_skips,
        pending_sends,
        connectable_yield,
        recommendation,
    }
}

pub(crate) fn source_yield_report(run: &Run) -> Vec<SourceYieldStats> {
    run.sources
        .iter()
        .map(|source| source_yield_stats(run, source))
        .collect()
}

pub(crate) fn low_yield_source_names(
    run: &Run,
    min_raw_rows: u32,
    max_connectable_yield: f64,
) -> Vec<String> {
    source_yield_report(run)
        .into_iter()
        .filter(|stats| stats.raw_row_count >= min_raw_rows)
        .filter(|stats| {
            stats
                .connectable_yield
                .is_some_and(|yield_rate| yield_rate <= max_connectable_yield)
        })
        .map(|stats| stats.source)
        .collect()
}

File: archive/legacy-go-js/internal/app/salesnav.go
package app

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

type SalesNavCapture struct {
	CapturedAt     *string               `json:"capturedAt"`
	Source         *string               `json:"source"`
	URL            *string               `json:"url"`
	ResumeURL      *string               `json:"resumeUrl"`
	Page           *SalesNavCapturePage  `json:"page"`
	Pages          []SalesNavCapturePage `json:"pages"`
	StateCounts    map[string]uint32     `json:"stateCounts"`
	RawRowCount    *uint32               `json:"rawRowCount"`
	OutputRowCount *uint32               `json:"outputRowCount"`
	Rows           []SalesNavCaptureRow  `json:"rows"`
}

type SalesNavCapturePage struct {
	URL       *string `json:"url"`
	PageLabel *string `json:"pageLabel"`
}

type ImportCaptureOptions struct {
	OnlyConnectable bool
}

type SalesNavAudit struct {
	PeopleCount uint32   `json:"peopleCount"`
	RecentNames []string `json:"recentNames"`
}

func (a *SalesNavAudit) UnmarshalJSON(data []byte) error {
	type primary struct {
		PeopleCount      *uint32  `json:"peopleCount"`
		PeopleCountSnake *uint32  `json:"people_count"`
		RecentNames      []string `json:"recentNames"`
		RecentNamesSnake []string `json:"recent_names"`
	}
	var value primary
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}
	if value.PeopleCount != nil {
		a.PeopleCount = *value.PeopleCount
	} else if value.PeopleCountSnake != nil {
		a.PeopleCount = *value.PeopleCountSnake
	}
	if value.RecentNames != nil {
		a.RecentNames = value.RecentNames
	} else if value.RecentNamesSnake != nil {
		a.RecentNames = value.RecentNamesSnake
	} else {
		a.RecentNames = []string{}
	}
	return nil
}

func LoadSalesNavAudit(path string) (SalesNavAudit, error) {
	var audit SalesNavAudit
	if err := readJSONFile(path, &audit, "reading audit "+path, "parsing audit "+path); err != nil {
		return SalesNavAudit{}, err
	}
	if audit.RecentNames == nil {
		audit.RecentNames = []string{}
	}
	return audit, nil
}

type SalesNavSendResult struct {
	Candidate SalesNavSendCandidate `json:"candidate"`
	Status    string                `json:"status"`
	Send      json.RawMessage       `json:"send"`
}

func LoadSalesNavSendResult(path string) (SalesNavSendResult, error) {
	var result SalesNavSendResult
	if err := readJSONFile(path, &result, "reading send "+path, "parsing send "+path); err != nil {
		return SalesNavSendResult{}, err
	}
	return result, nil
}

func (r SalesNavSendResult) ToCandidateStatus() (CandidateStatus, string) {
	switch r.Status {
	case "pending-verified":
		return CandidateStatusPending, "salesnav-send-one verified Connect - Pending"
	case "already-pending":
		return CandidateStatusAlreadyPending, "salesnav-send-one found already pending"
	case "email-required":
		return CandidateStatusSkipped, "salesnav-send-one stopped on email-required invite flow"
	default:
		send := "no send detail"
		if len(r.Send) > 0 && string(r.Send) != "null" {
			send = string(r.Send)
		}
		return CandidateStatusFailed, fmt.Sprintf("salesnav-send-one status %s; %s", r.Status, send)
	}
}

type SalesNavSendCandidate struct {
	Source     string  `json:"source"`
	Name       string  `json:"name"`
	ProfileURL *string `json:"profileUrl"`
}

func (c *SalesNavSendCandidate) UnmarshalJSON(data []byte) error {
	type primary struct {
		Source          string  `json:"source"`
		Name            string  `json:"name"`
		ProfileURL      *string `json:"profileUrl"`
		ProfileURLSnake *string `json:"profile_url"`
	}
	var value primary
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}
	c.Source = value.Source
	c.Name = value.Name
	c.ProfileURL = value.ProfileURL
	if c.ProfileURL == nil {
		c.ProfileURL = value.ProfileURLSnake
	}
	return nil
}

func LoadSalesNavCapture(path string) (SalesNavCapture, error) {
	var capture SalesNavCapture
	if err := readJSONFile(path, &capture, "reading capture "+path, "parsing capture "+path); err != nil {
		return SalesNavCapture{}, err
	}
	if capture.Pages == nil {
		capture.Pages = []SalesNavCapturePage{}
	}
	if capture.StateCounts == nil {
		capture.StateCounts = map[string]uint32{}
	}
	if capture.Rows == nil {
		capture.Rows = []SalesNavCaptureRow{}
	}
	return capture, nil
}

type SalesNavCaptureRow struct {
	Index        uint32                     `json:"index"`
	Name         *string                    `json:"name"`
	Text         *string                    `json:"text"`
	ProfileURL   *string                    `json:"profileUrl"`
	ScrollURN    *string                    `json:"scrollUrn"`
	VisibleState json.RawMessage            `json:"visibleState"`
	MenuState    *string                    `json:"menuState"`
	MenuLabels   []SalesNavCaptureMenuLabel `json:"menuLabels"`
	Links        []SalesNavCaptureLink      `json:"links"`
	RowHTMLPath  *string                    `json:"rowHtmlPath"`
}

type SalesNavCaptureMenuLabel struct {
	Text *string `json:"text"`
	Aria *string `json:"aria"`
}

type SalesNavCaptureLink struct {
	Text *string `json:"text"`
	Aria *string `json:"aria"`
	Href *string `json:"href"`
}

func CaptureStateCount(capture SalesNavCapture, state string) uint32 {
	if value, ok := capture.StateCounts[state]; ok {
		return value
	}
	var count uint32
	for _, row := range capture.Rows {
		if row.MenuState != nil && *row.MenuState == state {
			count++
		}
	}
	return count
}

func SameObservationIdentity(left, right CandidateObservation) bool {
	if left.ProfileURL != nil && right.ProfileURL != nil {
		return NormalizeLinkedInURL(*left.ProfileURL) == NormalizeLinkedInURL(*right.ProfileURL)
	}
	return left.Source == right.Source && left.Name == right.Name
}

func CandidateMatchesObservation(candidate CandidateEvent, observation CandidateObservation) bool {
	if candidate.ProfileURL != nil && observation.ProfileURL != nil {
		return NormalizeLinkedInURL(*candidate.ProfileURL) == NormalizeLinkedInURL(*observation.ProfileURL)
	}
	return candidate.Name == observation.Name && candidate.Source == observation.Source
}

func IsAutoStaleSkip(candidate CandidateEvent) bool {
	return candidate.Status == CandidateStatusSkipped && candidate.Note != nil && strings.Contains(*candidate.Note, "auto-skipped stale imported candidate")
}

func SalesProfileURNToLeadURL(urn string) *string {
	start := strings.Index(urn, "(")
	if start < 0 || !strings.HasSuffix(urn, ")") {
		return nil
	}
	tuple := strings.TrimSuffix(urn[start+1:], ")")
	parts := strings.Split(tuple, ",")
	if len(parts) != 3 {
		return nil
	}
	for i := range parts {
		parts[i] = strings.TrimSpace(parts[i])
		if parts[i] == "" {
			return nil
		}
	}
	value := fmt.Sprintf("https://www.linkedin.com/sales/lead/%s,%s,%s", parts[0], parts[1], parts[2])
	return &value
}

func PushTiming(run *Run, phase string, source *string, started time.Time, detail *string) {
	durationMS := uint64(time.Since(started).Milliseconds())
	run.Timings = append(run.Timings, RunTimingEvent{
		At:         time.Now(),
		Phase:      phase,
		Source:     source,
		DurationMS: durationMS,
		Detail:     detail,
	})
	run.MarkUpdated()
}

func CaptureToObservations(source string, capture SalesNavCapture, options ImportCaptureOptions) []CandidateObservation {
	observations := []CandidateObservation{}
	for _, row := range capture.Rows {
		if row.Name == nil || strings.TrimSpace(*row.Name) == "" {
			continue
		}
		menuState := "unknown"
		if row.MenuState != nil {
			menuState = *row.MenuState
		}
		if options.OnlyConnectable && menuState != "connectable" {
			continue
		}
		labels := []string{}
		for _, label := range row.MenuLabels {
			var value *string
			if label.Text != nil {
				value = label.Text
			} else {
				value = label.Aria
			}
			if value == nil {
				continue
			}
			cleaned := strings.TrimSpace(*value)
			if cleaned != "" {
				labels = append(labels, cleaned)
			}
		}
		profileURL := row.ProfileURL
		if profileURL == nil && row.ScrollURN != nil {
			profileURL = SalesProfileURNToLeadURL(*row.ScrollURN)
		}
		visibleState := row.VisibleState
		if len(visibleState) == 0 {
			visibleState = json.RawMessage("null")
		}
		observations = append(observations, CandidateObservation{
			ImportedAt:      time.Now(),
			CapturedAt:      capture.CapturedAt,
			Source:          source,
			Index:           row.Index,
			Name:            strings.TrimSpace(*row.Name),
			ProfileURL:      profileURL,
			SalesProfileURN: row.ScrollURN,
			VisibleState:    visibleState,
			MenuState:       menuState,
			MenuLabels:      labels,
			RowHTMLPath:     row.RowHTMLPath,
		})
	}
	return observations
}

func UpdateCaptureCursor(run *Run, source string, capture SalesNavCapture) {
	var lastPage *SalesNavCapturePage
	if capture.Page != nil {
		lastPage = capture.Page
	} else if len(capture.Pages) > 0 {
		lastPage = &capture.Pages[len(capture.Pages)-1]
	}
	resumeURL := capture.ResumeURL
	if resumeURL == nil {
		resumeURL = capture.URL
	}
	if resumeURL == nil && lastPage != nil {
		resumeURL = lastPage.URL
	}
	capturedPages := uint32(len(capture.Pages))
	if capturedPages == 0 && capture.Page != nil {
		capturedPages = 1
	}
	rawRowCount := uint32(len(capture.Rows))
	if capture.RawRowCount != nil {
		rawRowCount = *capture.RawRowCount
	}
	outputRowCount := uint32(len(capture.Rows))
	if capture.OutputRowCount != nil {
		outputRowCount = *capture.OutputRowCount
	}
	var pageLabel *string
	if lastPage != nil {
		pageLabel = lastPage.PageLabel
	}
	run.CaptureCursors[source] = SourceCaptureCursor{
		Source:              source,
		UpdatedAt:           time.Now(),
		CapturedAt:          capture.CapturedAt,
		ResumeURL:           resumeURL,
		PageLabel:           pageLabel,
		CapturedPages:       capturedPages,
		RawRowCount:         rawRowCount,
		OutputRowCount:      outputRowCount,
		ConnectableCount:    CaptureStateCount(capture, "connectable"),
		AlreadyPendingCount: CaptureStateCount(capture, "already-pending"),
		MissingTriggerCount: CaptureStateCount(capture, "missing-trigger"),
		StateCounts:         capture.StateCounts,
	}
}

func ImportCapture(run *Run, capture SalesNavCapture, options ImportCaptureOptions) (int, error) {
	source := ""
	if capture.Source != nil {
		source = *capture.Source
	} else if next := run.NextSource(); next != nil {
		source = next.Name
	} else {
		return 0, fmt.Errorf("capture did not include source and run has no next source")
	}
	if err := EnsureKnownSource(*run, source); err != nil {
		return 0, err
	}
	UpdateCaptureCursor(run, source, capture)
	imported := 0
	for _, observation := range CaptureToObservations(source, capture, options) {
		existingIndex := -1
		for i, existing := range run.Observations {
			if SameObservationIdentity(existing, observation) {
				existingIndex = i
				break
			}
		}
		if existingIndex >= 0 {
			run.Observations[existingIndex] = observation
		} else {
			run.Observations = append(run.Observations, observation)
			imported++
		}
	}
	return imported, nil
}

func ImportCaptureIntoReservoir(reservoir *CandidateReservoir, capture SalesNavCapture, options ImportCaptureOptions) (int, error) {
	if capture.Source == nil {
		return 0, fmt.Errorf("capture did not include source")
	}
	imported := 0
	for _, observation := range CaptureToObservations(*capture.Source, capture, options) {
		existingIndex := -1
		for i, existing := range reservoir.Observations {
			if SameObservationIdentity(existing, observation) {
				existingIndex = i
				break
			}
		}
		if existingIndex >= 0 {
			reservoir.Observations[existingIndex] = observation
		} else {
			reservoir.Observations = append(reservoir.Observations, observation)
			imported++
		}
	}
	now := time.Now()
	reservoir.UpdatedAt = &now
	return imported, nil
}

func FillRunFromReservoir(run *Run, reservoir *CandidateReservoir, source string, limit int) (int, error) {
	if err := EnsureKnownSource(*run, source); err != nil {
		return 0, err
	}
	selectedKeys := map[string]bool{}
	imported := 0
	for _, observation := range reservoir.AvailableForRunSource(*run, source) {
		if imported >= limit {
			break
		}
		observation.ImportedAt = time.Now()
		selectedKeys[ObservationKey(observation)] = true
		run.Observations = append(run.Observations, observation)
		imported++
	}
	if imported > 0 {
		remaining := reservoir.Observations[:0]
		for _, observation := range reservoir.Observations {
			if !selectedKeys[ObservationKey(observation)] {
				remaining = append(remaining, observation)
			}
		}
		reservoir.Observations = remaining
		now := time.Now()
		reservoir.UpdatedAt = &now
		run.MarkUpdated()
	}
	return imported, nil
}

func FillRunFromReservoirForTopUp(run *Run, reservoir *CandidateReservoir, source string, limit int) (int, error) {
	if err := EnsureKnownSource(*run, source); err != nil {
		return 0, err
	}
	selectedKeys := map[string]bool{}
	selected := []CandidateObservation{}
	for _, observation := range reservoir.Observations {
		if len(selected) >= limit {
			break
		}
		if observation.Source != source || observation.MenuState != "connectable" || run.HasTopUpBlockingEventForObservation(observation) {
			continue
		}
		exists := false
		for _, existing := range run.Observations {
			if SameObservationIdentity(existing, observation) {
				exists = true
				break
			}
		}
		if exists {
			continue
		}
		observation.ImportedAt = time.Now()
		selectedKeys[ObservationKey(observation)] = true
		selected = append(selected, observation)
	}
	if len(selected) > 0 {
		run.Observations = append(run.Observations, selected...)
		remaining := reservoir.Observations[:0]
		for _, observation := range reservoir.Observations {
			if !selectedKeys[ObservationKey(observation)] {
				remaining = append(remaining, observation)
			}
		}
		reservoir.Observations = remaining
		now := time.Now()
		reservoir.UpdatedAt = &now
		run.MarkUpdated()
	}
	return len(selected), nil
}

func ObservationKey(observation CandidateObservation) string {
	profile := ""
	if observation.ProfileURL != nil {
		profile = NormalizeLinkedInURL(*observation.ProfileURL)
	}
	return observation.Source + "\x00" + observation.Name + "\x00" + profile
}

func SourceYieldStatsForRun(run Run, source SourcePlan) SourceYieldStats {
	cursor, hasCursor := run.CaptureCursors[source.Name]
	rawRowCount := uint32(0)
	connectableCount := uint32(0)
	alreadyPendingCount := uint32(0)
	if hasCursor {
		rawRowCount = cursor.RawRowCount
		connectableCount = cursor.ConnectableCount
		alreadyPendingCount = cursor.AlreadyPendingCount
	} else {
		for _, observation := range run.Observations {
			if observation.Source == source.Name && observation.MenuState == "connectable" {
				connectableCount++
			}
		}
	}
	var emailRequiredSkips uint32
	for _, candidate := range run.Candidates {
		if candidate.Source != source.Name || candidate.Status != CandidateStatusSkipped || candidate.Note == nil {
			continue
		}
		if strings.Contains(strings.ToLower(*candidate.Note), "email-required") {
			emailRequiredSkips++
		}
	}
	pendingSends := run.SourceVerifiedCount(source.Name)
	var yield *float64
	if rawRowCount > 0 {
		value := float64(connectableCount) / float64(rawRowCount)
		yield = &value
	}
	recommendation := "no capture data"
	if yield != nil {
		switch {
		case rawRowCount >= 50 && *yield <= 0.05:
			recommendation = "low-yield: consider reservoir/fallback before deeper capture"
		case rawRowCount >= 25 && *yield <= 0.10:
			recommendation = "thin: capture with a small buffer and be ready to carry over"
		default:
			recommendation = "ok"
		}
	}
	return SourceYieldStats{
		Source:              source.Name,
		RawRowCount:         rawRowCount,
		ConnectableCount:    connectableCount,
		AlreadyPendingCount: alreadyPendingCount,
		EmailRequiredSkips:  emailRequiredSkips,
		PendingSends:        pendingSends,
		ConnectableYield:    yield,
		Recommendation:      recommendation,
	}
}

func SourceYieldReport(run Run) []SourceYieldStats {
	result := make([]SourceYieldStats, 0, len(run.Sources))
	for _, source := range run.Sources {
		result = append(result, SourceYieldStatsForRun(run, source))
	}
	return result
}

func LowYieldSourceNames(run Run, minRawRows uint32, maxConnectableYield float64) []string {
	names := []string{}
	for _, stats := range SourceYieldReport(run) {
		if stats.RawRowCount >= minRawRows && stats.ConnectableYield != nil && *stats.ConnectableYield <= maxConnectableYield {
			names = append(names, stats.Source)
		}
	}
	return names
}

func SaveRawJSON(path string, value any) error {
	raw, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return os.WriteFile(path, raw, 0o644)
}
